### ISSUE TITLE
Possible "conan config get storage.path" regression

### ISSUE THREAD

**Issue by jcar87** [conan-io/conan#5338]
On a clean install of conan, starting from version 1.15.0:

```
$ conan config get storage.path
./data
```

On older versions: 
```
$ conan config get storage.path
~/.conan/data
```

I suspect this affects new installs and not upgrades.  To reproduce:
```
rm -rf ~/.conan
pip install -I conan==1.15.0
conan config get storage.path
rm -rf ~/.conan
pip install -I conan==1.14.5
```

Saludos!

**Comment by memsharded** [comment]
Yes, the default path in the ``[storage]`` section has changed its default representation, but it still resolves to the same folder, right? Or is that failing in some case? The new ``./data`` representation is relative to the ``conan.conf`` location, which is more intutive for most cases, and removes the annoying ``.conan`` intermediate folder. But using the old one should totally work, please tell if it doesn't.

**Comment by nolange** [comment]
How would one get the absolute path then?

I currently use the following commands, because its otherwise not possible to use conan with a properly isolated build environment (#4967).
I would need the ideally fully resolved path to the data directory.

```cmake
execute_process(COMMAND conan config get storage.path OUTPUT_VARIABLE CMAKE_CONAN_STORAGE_PATH OUTPUT_STRIP_TRAILING_WHITESPACE)
set(CMAKE_STAGING_PREFIX "${CMAKE_CONAN_STORAGE_PATH}")

```

**Comment by memsharded** [comment]
Hi @nolange 

I am not fully sure what is the purpose of setting this path for CMAKE_STAGING_PREFIX. I have tried reading the docs for that var, but not very informative. Also in internet, but still can't fully understand. What if you directly set?
```cmake
set(CMAKE_STAGING_PREFIX "~")
```
Assuming that cmake knows how to expand ``~`` as the userhome, otherwise, it will not work anyway. Does cmake expand correctly ``~``?

What would be the difference of setting up that or:
```cmake
set(CMAKE_STAGING_PREFIX "~/.conan")
# or
set(CMAKE_STAGING_PREFIX "~/.conan/data")
```

Also, what happens to packages which final path is not really there, but mapped elsewhere, like when using ``short_paths``?

Needing to setup something in the build system that points to the storage path sounds like it could be problematic. I understand the need to restrict cmake to NOT look in other paths than those given by the conan generators, but can't fully understand this CMAKE_STAGING_PREFIX variable.

It might be possible to return the fully evaluated path for ``storage.path`` field, but sounds a bit hacky, I am investigating other possibilities.

**Comment by jcar87** [comment]
> Yes, the default path in the `[storage]` section has changed its default representation, but it still resolves to the same folder, right? Or is that failing in some case? The new `./data` representation is relative to the `conan.conf` location, which is more intutive for most cases, and removes the annoying `.conan` intermediate folder. But using the old one should totally work, please tell if it doesn't.

Yup, as far as `conan.conf` is concerned, I think the new layout makes sense as it is clearer now where it is relative to.  

My specific issue has more to do with the output of  `conan config get storage.path` itself: prior to this version, I feel like it could be used to retrieve and compute the path programmatically, e.g. `~/.conan/data` can be expanded to the full path depending on whether `CONAN_USER_HOME` is set (I could be mistaken, I dont use `CONAN_USER_HOME` very often).

Whereas, an output like  `./data` seems to be lacking some information. From an outside perspective, would it be safe to assume that the path is relative to `<HOME>/.conan/`  (depending on whether `CONAN_USER_HOME` is overriding the user home? That would work for my use case.

On the other hand, it would seem that the output of `conan config get storage.path` would need to be handled differently depending on the conan version.

**Comment by nolange** [comment]
This is getting a bit off-topic, the linked bug has more information, as well as the documentation for `find_package`.

I have a cross-toolchain, which also sets those variables (exemplary values):

```cmake
set(CMAKE_SYSROOT /opt/customroot)
# thats ensuring you dont end up with a system-installed package
set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ONLY)
```

and a path set by the generated `conan_paths.cmake`

```cmake
set(CMAKE_PREFIX_PATH "/home/me/.conan/data/gtest...." ${CMAKE_PREFIX_PATH} ${CMAKE_CURRENT_LIST_DIR})
```

What will happen, is that `find_package` will re-root the prefix under `/opt/customroot`, and thus search in `/opt/customroot/home/me/.conan/data/gtest....`. Naturally there is nothing to be found.

the directory in `CMAKE_STAGING_PREFIX` will be excluded from the re-rooting, hence the need to set it to the data directory.

> What if you directly set?
> 
> ```cmake
> set(CMAKE_STAGING_PREFIX "~")
> ```
> 
> Assuming that cmake knows how to expand `~` as the userhome, otherwise, it will not work anyway. Does cmake expand correctly `~`?

Then I am hard-coding assumptions about conan and its directories, likely need different values across Linux/Windows/Mac, and its not restrictive as it could be. CMake cant resolve `~` correctly under all circumstances, as this is affected by `CONAN_USER_HOME`.

> Also, what happens to packages which final path is not really there, but mapped elsewhere, like when using `short_paths`?

I would like conan to give me the relevant path (thats the whole idea behind not hard-coding)

> Needing to setup something in the build system that points to the storage path sounds like it could be problematic. I understand the need to restrict cmake to NOT look in other paths than those given by the conan generators, but can't fully understand this CMAKE_STAGING_PREFIX variable.

Yeah, thats why I opened issue 4967 to figure out a good solution, until then this is the best I could come up with (now it's broken with a new default config)

> It might be possible to return the fully evaluated path for `storage.path` field, but sounds a bit hacky, I am investigating other possibilities.

I would prefer conan to set a variable like `CONAN_STORAGE_PATH` in the file `conan_paths.cmake`

**Comment by memsharded** [comment]
I am proposing in #5350 to return the full, resolved path, which seems to make sense, specially because the real configuration can be different if env-vars like CONAN_STORAGE_PATH is defined, in which case, the thing that was previously being returned is useless and out-of-sync with the real value that is applied when conan runs.

**Comment by memsharded** [comment]
Merged, to be released in Conan 1.16.1
